include/eosio/opaque.hpp

```diff
@@ -0,0 +1,134 @@
+/// thus saving some CPU cycles or memory requirement. 
+///
+/// For example, given a foo_type,
+/// 
+/// <code>
+///    struct foo_type {
+///      uint32_t                    field1;
+///      string                      field2;
+///      opaque<std::vector<string>> filed3;
+///    };
+/// </code>
+///   
+/// the deserialization can be implemented as follows:
+/// 
+/// <code>  
+///   input_stream serialized_foo_stream(...);
+///   foo_type foo_value; 
+///   from_bin(foo_value, serialized_foo_stream);
+///   if (foo_value.field1 > 1 || foo_value.field2 == "meet_precondition") {
+///      auto sz = foo_value.field3.unpack_size();
+///      for (size_t i = 0; i < sz; ++i) {
+///         if (foo_value.field3.unpack() == "value_to_be_searched") {
+///            do_something_when_found();
+///            break;
+///         }
+///      }
+///   }
+/// </code>
+
+template <typename T>
+class opaque_base {
+ protected:
+   input_stream bin;
+
+ public:
+   opaque_base() = default;
+   explicit opaque_base(const std::vector<char>& data) : bin(data) {}
+   explicit opaque_base(input_stream strm) { eosio::from_bin(bin, strm); }
+
+   void unpack(T& obj) { eosio::from_bin(obj, bin); }
+
+   T unpack() {
+      T obj;
+      this->unpack(obj);
+      return obj;
+   }
+
+   bool empty() const { return !bin.remaining(); }
+
+   template <typename S>
+   void from(S& stream) {
+      eosio::from_bin(this->bin, stream);
+   }
+};
+
+template <typename T>
+class opaque : public opaque_base<T> {
+ public:
+   using opaque_base<T>::opaque_base;
+};
+
+template <typename T>
+class opaque<std::vector<T>> : public opaque_base<std::vector<T>> {
+ public:
+   using opaque_base<std::vector<T>>::opaque_base;
+
+   size_t unpack_size() {
+      uint32_t num;
+      varuint32_from_bin(num, this->bin);
+      return num;
+   }
+
+   void unpack_next(T& obj) { eosio::from_bin(obj, this->bin); }
+
+   T unpack_next() {
+      T obj;
+      this->unpack_next(obj);
+      return obj;
+   }
+};
+
+template <typename T>
+constexpr auto get_opaque_type_name() {
+   constexpr std::size_t  size = sizeof("opaque") + (std::string_view(get_type_name((T*)nullptr)).size() + 1);
+   std::array<char, size> buffer{ 'o', 'p', 'a', 'q', 'u', 'e' };
+   (variant_type_appender{ buffer.data() + 6 } + std::string_view(get_type_name((T*)nullptr)));
+   buffer[buffer.size() - 1] = '\0';
+   return buffer;
+}
+
+template <typename T>
+constexpr auto get_opaque_vector_type_name() {
+   constexpr std::size_t size =
+         sizeof("opaque") + std::string_view(get_type_name((T*)nullptr)).size() + sizeof("vector") + 1;
+   std::array<char, size> buffer{ 'o', 'p', 'a', 'q', 'u', 'e' };
+   (variant_type_appender{ buffer.data() + 6 } + std::string_view(get_type_name((T*)nullptr)) +
+    std::string_view("vector"));
+   buffer[buffer.size() - 1] = '\0';
+   return buffer;
+}
+
+template <typename T>
+constexpr auto opaque_type_name = get_opaque_type_name<T>();
```

Comment:
Is this actually necessary?  The type name is used for generating an ABI, and the ABI type for all opaque types is bytes.